### ISSUE TITLE
Excerpt by word count

### ISSUE THREAD

**Issue by jrwarwick** [OraOpenSource/oos-utils#5]
```SQL
-- SQL inline version:
-- The target word count is indicated in two places, and should agree: the second value in the curly-braces range specification in the call to regexp_substr, and then the first (and only) value in the call to regexp_instr 
select REGEXP_SUBSTR( QT.LONG_DESCRIPTION, '((\w*\W){1,8})') || DECODE(SIGN(REGEXP_INSTR(QT.LONG_DESCRIPTION, '((\w*\W){8,})')),1,'...') EXCERPT,
  qt.long_description 
from ( select 'Some really, really long piece of text with many words in it. Sometimes you do not want to just chop the text and you also want a nice and neat portion to display.' long_description from dual ) qt
;


-- as a nice pl/sql function:

/**
 * Return a conventional English ellipsis-indicated excerptation of a longer text.
 *
 * @author Justin Warwick
 * @created 2015-04-24
 * @param p_long_description raw longer text
 * @param p_max_count target word count, but it is flexible. 
 *        If the text is too short, no abbreviation is performed. 
 *        If a negative word count is supplied, treat it as the number of words to leave off the end.
 * @return varchar2 formatted excerpt
 */

create or replace function excerpt_by_wordcount(p_long_description in varchar2, p_max_count in number := 8) return varchar2 as
  l_word_count number;
  l_max_count number;
BEGIN
  if p_max_count < 0 then
    l_word_count := regexp_count(p_long_description,'(\w*\W*( |$))');
    l_max_count := l_word_count + p_max_count;
    if (l_max_count < 1) then
      l_max_count := l_word_count;
    end if;
  else
    l_max_count := p_max_count;
  end if;
    
  return REGEXP_SUBSTR( p_long_description, '((\w*\W){1,'||l_max_count||'})') || CASE SIGN(REGEXP_INSTR(p_long_description, '((\w*\W){'||l_max_count||',})')) WHEN 1 THEN '...' END;  
END
;
/

```

**Comment by martindsouza** [comment]
I decided to take the idea and modify it a bit. I created a function called `truncate_string` which can truncate by length or by word and append an ellipsis to it. 

The length that is provided is the guaranteed maximum length returned by the function. This is different than the goal of your proposal. It is also a bit more complex as we can't do lookbehind assertions in Oracle.

Let me know what you think.

**Comment by jrwarwick** [comment]
Sorry for the late response. I think your changes make good sense.
Thanks,

  --Justin Warwick


I decided to take the idea and modify it a bit. I created a function called
truncate_string which can truncate by length or by word and append an
ellipsis to it.

The length that is provided is the guaranteed maximum length returned by
the function. This is different than the goal of your proposal. It is also
a bit more complex as we can't do lookbehind assertions in Oracle.

Let me know what you think.

—
Reply to this email directly or view it on GitHub
<https://github.com/OraOpenSource/oos-utils/issues/5#issuecomment-138396609>
.
